cmd/ambex/main.go

```diff
@@ -643,77 +627,110 @@ func update(ctx context.Context, snapdirPath string, numsnaps int, config v2cach
 	}
 }
 
-func warn(err error) bool {
+func warn(ctx context.Context, err error) bool {
 	if err != nil {
-		log.Warn(err)
+		dlog.Warn(ctx, err)
 		return true
 	} else {
 		return false
 	}
 }
 
-// OnStreamOpen is called once an xDS stream is open with a stream ID and the type URL (or "" for ADS).
-func (l loggerv2) OnStreamOpen(_ context.Context, sid int64, stype string) error {
-	l.Debugf("V2 Stream open[%v]: %v", sid, stype)
-	return nil
+type logAdapterBase struct {
+	prefix string
 }
 
-func (l loggerv3) OnStreamOpen(_ context.Context, sid int64, stype string) error {
-	l.Debugf("V3 Stream open[%v]: %v", sid, stype)
-	return nil
+type logAdapterV2 struct {
+	logAdapterBase
 }
 
-// OnStreamClosed is called immediately prior to closing an xDS stream with a stream ID.
-func (l loggerv2) OnStreamClosed(sid int64) {
-	l.Debugf("V2 Stream closed[%v]", sid)
+var _ v2server.Callbacks = logAdapterV2{}
+var _ envoyLog.Logger = logAdapterV2{}
+
+type logAdapterV3 struct {
+	logAdapterBase
+}
+
+var _ v3server.Callbacks = logAdapterV3{}
+var _ envoyLog.Logger = logAdapterV3{}
+
+// Debugf implements envoyLog.Logger.
```

Comment:
"XMethod implements YInterface." is a pretty common thing to write for a YInterface that specifies an XMethod method.  Sure, if YInterface is more than one method then XMethod doesn't implement YInterface without the other methods helping too.  But that seems nit-picky to me.

Unless you're saying that it's confusing because `logAdapterBase` doesn't implement any of it; only providing assistance to the `logAdapterV2`/`3`.  (I don't think that's actually true for `envoyLog.Logger`, but is true for `vXserver.Callbacks`).  I guess that's a fair point, but the code is all right there for you to see what's going on.  If it were more LOC or spread through multiple files, I'd agree with you.